src/main/java/net/minecraftforge/common/crafting/DifferenceIngredient.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Minecraft Forge - Forge Development LLC
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.common.crafting;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntComparators;
+import it.unimi.dsi.fastutil.ints.IntList;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.world.entity.player.StackedContents;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+
+import javax.annotation.Nullable;
+import java.util.Arrays;
+
+/** Ingredient that matches everything from the first ingredient that is not included in the second ingredient */
+public class DifferenceIngredient extends AbstractIngredient
+{
+    private final Ingredient base;
+    private final Ingredient subtracted;
+    private ItemStack[] filteredMatchingStacks;
+    private IntList packedMatchingStacks;
+
+    protected DifferenceIngredient(Ingredient base, Ingredient subtracted)
+    {
+        this.base = base;
+        this.subtracted = subtracted;
+    }
+
+    /**
+     * Gets the difference from the two ingredients
+     * @param base        Ingredient the item must match
+     * @param subtracted  Ingredient the item must not match
+     * @return  Ingredient that {@code base} anything in base that is not in {@code subtracted}
+     */
+    public static DifferenceIngredient of(Ingredient base, Ingredient subtracted)
+    {
+        return new DifferenceIngredient(base, subtracted);
+    }
+
+    @Override
+    public boolean test(@Nullable ItemStack stack)
+    {
+        if (stack == null || stack.isEmpty())
+            return false;
+        return base.test(stack) && !subtracted.test(stack);
+    }
+
+    @Override
+    public ItemStack[] getItems()
+    {
+        if (this.filteredMatchingStacks == null)
+            this.filteredMatchingStacks = Arrays.stream(base.getItems())
+						                                    .filter(stack -> !subtracted.test(stack))
```

Comment:
Weirdly out of line.

Reply:
Seems I need to argue with my formatter more, this is due to github's "tabs are 8 characters wide"